server/tables/tables.go

```diff
@@ -563,6 +577,61 @@ func PreAutoMigrate(db *gorm.DB) ([]PostAutoMigrateLogic, error) {
 			return nil
 		})
 	}
+
+	// Populate invocation_uuid if the column doesn't exist.
+	if m.HasTable("Invocations") && !m.HasColumn(&Invocation{}, "invocation_uuid") {
+		if db.Dialector.Name() == sqliteDialect {
+			// Close the rows before executing update statements for SQLite; otherwise, it will result in a deadlock.
+			postMigrate = append(postMigrate, func() error {
+				rows, err := db.Raw(`SELECT invocation_id, invocation_pk FROM Invocations`).Rows()
+				if err != nil {
+					return err
+				}
+				ids := []invocationIDs{}
+				for rows.Next() {
+					var invocationID string
+					var invocationPK int64
+					if err := rows.Scan(&invocationID, &invocationPK); err != nil {
+						return err
+					}
+					invocationUUID, err := uuid.StringToBytes(invocationID)
+					if err != nil {
+						return err
+					}
+					ids = append(ids, invocationIDs{
+						invocationID:   invocationID,
+						invocationUUID: invocationUUID,
+						invocationPK:   invocationPK,
+					})
+				}
+				rows.Close()
+
+				for _, id := range ids {
+					if err := db.Exec(`UPDATE TargetStatuses SET invocation_uuid = ? WHERE invocation_pk = ? `, id.invocationUUID, id.invocationPK).Error; err != nil {
+						return err
+					}
+					if err := db.Exec(`UPDATE Invocations SET invocation_uuid = ? WHERE invocation_id = ? `, id.invocationUUID, id.invocationID).Error; err != nil {
+						return err
+					}
+				}
+				return nil
+			})
+		} else if db.Dialector.Name() == mysqlDialect {
+			postMigrate = append(postMigrate, func() error {
+				if err := db.Exec(`UPDATE Invocations SET invocation_uuid = UUID_TO_BIN(invocation_id)`).Error; err != nil {
+					return err
+				}
```

Comment:
Tested locally with a fake prod db. It took ~17 min to finish updating Invocations; and ~42 mins to finish updating TargetStatuses. The whole migration took 1hr and 4 min locally.

Reply:
👍  thanks for testing the performance of this! I know how painful it can be waiting for these things to finish.

What's the plan for rolling this out -- how do you imagine we'll actually apply this update and how do we ensure that rows written during the schema migration also get migrated?

Reply:
I added a pre-migrate check. if the column exists, update invocation uuid if the column is null. The roll will look like this:
1. The first time the migration runs, it will create columns, indices and back fill invocation_uuids. During this time, other servers on the old versions might write the rows without invocation_uuids.
2. Other servers restarted. Then will go through the new migration code I just added; and updated invocation_uuids if they are null.

Reply:
I think with the code as written, the migration will run only once, since the migration is based on `!m.HasColumn(&Invocation{}, "invocation_uuid")` and the column will be created after the first new server instance completes its migration.

One thing you could do is replace that condition with (pseudo-code) `!m.HasColumn(&Invocation{}, "invocation_uuid") || query(SELECT COUNT(*) FROM Invocations WHERE invocation_uuid = NULL) > 0` -- as long as the COUNT runs very fast in the case where the migration is already complete. It'd probably be very fast since it'll see that there are no NULLs in the index, but worth verifying

Reply:
Nevermind, I missed the above case which does exactly this, lol.